pkg/resource/prunercronjob.go

```diff
@@ -0,0 +1,208 @@
+	"github.com/openshift/cluster-image-registry-operator/pkg/parameters"
+)
+
+const (
+	cronJobOwnerAnnotation = "imageregistry.openshift.io"
+)
+
+var (
+	defaultSuspend                          = false
+	defaultSchedule                         = "*/1 * * * *"
+	defaultStartingDeadlineSeconds    int64 = 60
+	defaultFailedJobsHistoryLimit     int32 = 3
+	defaultSuccessfulJobsHistoryLimit int32 = 3
+	defaultKeepTagRevisions                 = 3
+	defaultKeepYoungerThan                  = "60m"
+	defaultResources                        = kcorev1.ResourceRequirements{
+		Requests: kcorev1.ResourceList{
+			kcorev1.ResourceCPU:    resource.MustParse("1"),
+			kcorev1.ResourceMemory: resource.MustParse("1Gi"),
+		},
+	}
+)
+
+// PrunerCronJobIsCreatedByOperator returns whether or not the PrunerCronJob was created by the Operator
+func PrunerCronJobIsCreatedByOperator(prunerCronJob *batchapi.CronJob) bool {
+	_, ok := prunerCronJob.Annotations[cronJobOwnerAnnotation]
+	return ok
+}
+
+var _ Mutator = &generatorPrunerCronJob{}
+
+type generatorPrunerCronJob struct {
+	lister    batchlisters.CronJobNamespaceLister
+	client    batchset.BatchV1beta1Interface
+	namespace string
+	cr        *pruningapiv1.Config
+}
+
+func newGeneratorPrunerCronJob(lister batchlisters.CronJobNamespaceLister, client batchset.BatchV1beta1Interface, params *parameters.Globals, cr *pruningapiv1.Config) *generatorPrunerCronJob {
+	return &generatorPrunerCronJob{
+		lister: lister,
+		client: client,
+		cr:     cr,
+	}
+}
+
+func (gcj *generatorPrunerCronJob) Type() runtime.Object {
+	return &batchapi.CronJob{}
+}
+
+func (gcj *generatorPrunerCronJob) GetGroup() string {
+	return batchapi.GroupName
+}
+
+func (gcj *generatorPrunerCronJob) GetResource() string {
+	return "batches"
+}
+
+func (gcj *generatorPrunerCronJob) GetNamespace() string {
+	return gcj.namespace
+}
+
+func (gcj *generatorPrunerCronJob) GetName() string {
+	return "image-pruner"
+}
+
+func (gcj *generatorPrunerCronJob) expected() (runtime.Object, error) {
+	cj := &batchapi.CronJob{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        gcj.GetName(),
+			Namespace:   gcj.GetNamespace(),
+			Annotations: map[string]string{cronJobOwnerAnnotation: "true"},
+		},
+		Spec: batchapi.CronJobSpec{
+			Suspend:                    gcj.getSuspend(),
+			Schedule:                   gcj.getSchedule(),
+			ConcurrencyPolicy:          batchapi.ForbidConcurrent,
+			FailedJobsHistoryLimit:     gcj.getFailedJobsHistoryLimit(),
+			SuccessfulJobsHistoryLimit: gcj.getSuccessfulJobsHistoryLimit(),
+			StartingDeadlineSeconds:    gcj.getStartingDeadlineSeconds(),
+			JobTemplate: batchapi.JobTemplateSpec{
+				Spec: batchv1.JobSpec{
+					Template: kcorev1.PodTemplateSpec{
+						Spec: kcorev1.PodSpec{
+							RestartPolicy:      kcorev1.RestartPolicyOnFailure,
+							ServiceAccountName: "pruner",
+							Containers: []kcorev1.Container{
+								{
+									Image:                    "quay.io/openshift/origin-cli:4.1",
```

Comment:
@coreydaley I believe there is an `ImageStreamTag` for `oc` that we should reference instead. @gabemontero I believe the samples operator (or CVO) does this along with the must-gather tool?

If we have such imagestreamtag, we'd have to dynamically inject the pull spec to the internal registry - I don't think `CronJob` supports the image-lookup [1], and even so will it resolve imagestreams in the openshift namespace?

[1] https://docs.openshift.com/container-platform/3.11/admin_guide/image_policy.html#image-policy-configuring-the-image-policy-admission-plug-in

Reply:
+1, use the imagestream. this would also fail in disconnected envs and using a version tag that needs to be updated would be problematic.